### PR TITLE
[Modules] New module Azure Digital Twins and child endpoint

### PR DESCRIPTION
**Pull request by ryanmstephens**
# Description

New module Azure Digital Twins and child modules Azure Digital Twins endpoint. 

[Build Status](https://dev.azure.com/carml/CARML/_build/latest?definitionId=135&branchName=users%2Fryanmstephens%2F2135_digTw)

# Type of Change

Please delete options that are not relevant.

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Update to documentation

# Checklist

- [X] I'm sure there are no other open Pull Requests for the same update/change
- [X] My corresponding pipelines / checks run clean and green without any errors or warnings
- [X] My code follows the style guidelines of this project
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation (readme)
- [X] I did format my code

### CODE REVIEW

File: modules/Microsoft.DigitalTwins/digitalTwinsInstances/readme.md [hunk 1]
```diff
@@ -0,0 +1,612 @@
...
+<p>
+
+### Parameter Usage: `serviceBusEndpoint`
+
+<details>
+
+<summary>Parameter JSON format</summary>
+
+```json
+        "serviceBusEndpoint": {
```
Comment (AlexanderSehr) [inline modules/Microsoft.DigitalTwins/digitalTwinsInstances/readme.md:384]: Please update the identation of this and the other parameter usage examples

Comment (ryanmstephens) [reply modules/Microsoft.DigitalTwins/digitalTwinsInstances/readme.md]: Hi @AlexanderSehr

Good spot - Fixed in the latest commit.

File: modules/Microsoft.DigitalTwins/digitalTwinsInstances/.test/common/dependencies.bicep [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+
+output eventGridTopicName string = eventGridTopic.name
+
+output eventGridEndpoint string = eventGridDomain.properties.endpoint
+
+output eventGridTopicId string = eventGridTopic.id
+
+output eventGridDomainId string = eventGridDomain.id
+
+output managedIdentityId string = managedIdentity.id
```
Comment (AlexanderSehr) [inline modules/Microsoft.DigitalTwins/digitalTwinsInstances/.test/common/dependencies.bicep:166]: Each time you use `id` please use `resourceId` instead and add the output metadata. 
It doesn't provide a functional advantage, but makes it easier to understand what one is looking at and leads by example :)

File: modules/Microsoft.DigitalTwins/digitalTwinsInstances/.test/common/dependencies.bicep [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+        zoneRedundant: false
+        isAutoInflateEnabled: false
+        maximumThroughputUnits: 0
+    }
+}
+
+resource eventHub 'Microsoft.EventHub/namespaces/eventhubs@2022-10-01-preview' = {
+    name: eventHubName
+    parent: eventHubNamespace
+}
```
Comment (AlexanderSehr) [inline modules/Microsoft.DigitalTwins/digitalTwinsInstances/.test/common/dependencies.bicep:89]:
```suggestion
resource eventHubNamespace 'Microsoft.EventHub/namespaces@2022-10-01-preview' = {
    name: eventHubNamespaceName
    location: location
    properties: {
        zoneRedundant: false
        isAutoInflateEnabled: false
        maximumThroughputUnits: 0
    }

    resource eventHub 'eventhubs@2022-10-01-preview' = {
        name: eventHubName
    }
}
```

File: modules/Microsoft.DigitalTwins/digitalTwinsInstances/.test/common/dependencies.bicep [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+    location: location
+    properties: {
+        zoneRedundant: false
+    }
+}
+
+resource serviceBusTopic 'Microsoft.ServiceBus/namespaces/topics@2022-10-01-preview' = {
+    name: serviceBusTopicName
+    parent: serviceBus
+}
```
Comment (AlexanderSehr) [inline modules/Microsoft.DigitalTwins/digitalTwinsInstances/.test/common/dependencies.bicep:102]:
```suggestion
resource serviceBus 'Microsoft.ServiceBus/namespaces@2022-10-01-preview' = {
    name: serviceBusName
    location: location
    properties: {
        zoneRedundant: false
    }

    resource topic 'topics@2022-10-01-preview' = {
        name: serviceBusTopicName
    }
}
```

File: modules/Microsoft.DigitalTwins/digitalTwinsInstances/.test/common/dependencies.bicep [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+    location: location
+    properties: {
+        disableLocalAuth: false
+    }
+}
+
+resource eventGridTopic 'Microsoft.EventGrid/domains/topics@2022-06-15' = {
+    name: eventGridTopicName
+    parent: eventGridDomain
+}
```
Comment (AlexanderSehr) [inline modules/Microsoft.DigitalTwins/digitalTwinsInstances/.test/common/dependencies.bicep:115]:
```suggestion
resource eventGridDomain 'Microsoft.EventGrid/domains@2022-06-15' = {
    name: eventGridDomainName
    location: location
    properties: {
        disableLocalAuth: false
    }

    resource topic 'topics@2022-06-15' = {
        name: eventGridTopicName
    }
}
```

File: modules/Microsoft.DigitalTwins/digitalTwinsInstances/.test/common/dependencies.bicep [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+
+resource evhrbacAssignment 'Microsoft.Authorization/roleAssignments@2022-04-01' = {
+    name: guid(managedIdentity.id, 'evhrbacAssignment')
+    scope: eventHubNamespace
+    properties: {
+        roleDefinitionId: subscriptionResourceId('Microsoft.Authorization/roleDefinitions', '2b629674-e913-4c01-ae53-ef4638d8f975')
```
Comment (AlexanderSehr) [inline modules/Microsoft.DigitalTwins/digitalTwinsInstances/.test/common/dependencies.bicep:121]: Please add a comment in the back what role this is

File: modules/Microsoft.DigitalTwins/digitalTwinsInstances/.test/common/dependencies.bicep [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+        disableLocalAuth: false
+    }
+}
+
+resource eventGridTopic 'Microsoft.EventGrid/domains/topics@2022-06-15' = {
+    name: eventGridTopicName
+    parent: eventGridDomain
+}
+
+resource evhrbacAssignment 'Microsoft.Authorization/roleAssignments@2022-04-01' = {
```
Comment (AlexanderSehr) [inline modules/Microsoft.DigitalTwins/digitalTwinsInstances/.test/common/dependencies.bicep:117]:
```suggestion
resource eventHubNamespaceRbacAssignment 'Microsoft.Authorization/roleAssignments@2022-04-01' = {
```

File: modules/Microsoft.DigitalTwins/digitalTwinsInstances/.test/common/dependencies.bicep [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+    name: guid(managedIdentity.id, 'evhrbacAssignment')
+    scope: eventHubNamespace
+    properties: {
+        roleDefinitionId: subscriptionResourceId('Microsoft.Authorization/roleDefinitions', '2b629674-e913-4c01-ae53-ef4638d8f975')
+        principalId: managedIdentity.properties.principalId
+        principalType: 'ServicePrincipal'
+    }
+}
+
+resource sbhrbacAssignment 'Microsoft.Authorization/roleAssignments@2022-04-01' = {
```
Comment (AlexanderSehr) [inline modules/Microsoft.DigitalTwins/digitalTwinsInstances/.test/common/dependencies.bicep:127]:
```suggestion
resource serviceBusRbacAssignment 'Microsoft.Authorization/roleAssignments@2022-04-01' = {
```

File: modules/Microsoft.DigitalTwins/digitalTwinsInstances/.test/common/dependencies.bicep [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+    name: guid(managedIdentity.id, 'sbrbacAssignment')
+    scope: serviceBus
+    properties: {
+        roleDefinitionId: subscriptionResourceId('Microsoft.Authorization/roleDefinitions', '69a216fc-b8fb-44d8-bc22-1f3c2cd27a39')
```
Comment (AlexanderSehr) [inline modules/Microsoft.DigitalTwins/digitalTwinsInstances/.test/common/dependencies.bicep:131]: Please add a comment in the back what this role is

File: modules/Microsoft.DigitalTwins/digitalTwinsInstances/.test/common/dependencies.bicep [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+    }
+}
+
+resource managedIdentity 'Microsoft.ManagedIdentity/userAssignedIdentities@2018-11-30' = {
+    name: managedIdentityName
+    location: location
+}
+
+resource privateDNSZone 'Microsoft.Network/privateDnsZones@2020-06-01' = {
+    name: 'privatelink.vaultcore.azure.net'
```
Comment (AlexanderSehr) [inline modules/Microsoft.DigitalTwins/digitalTwinsInstances/.test/common/dependencies.bicep:61]: Is this really `vaultcore`? It's an honest question, but I'd have assumed it's something special for Digital Twins.

Comment (AlexanderSehr) [reply modules/Microsoft.DigitalTwins/digitalTwinsInstances/.test/common/dependencies.bicep:61]: like [`privatelink.digitaltwins.azure.net`](https://learn.microsoft.com/en-us/azure/digital-twins/how-to-enable-private-link?tabs=portal%2Cportal-2#add-a-private-endpoint-to-an-existing-instance)

File: modules/Microsoft.DigitalTwins/digitalTwinsInstances/.test/common/deploy.test.bicep [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+module nestedDependencies 'dependencies.bicep' = {
+  scope: resourceGroup
+  name: '${uniqueString(deployment().name, location)}-nestedDependencies'
+  params: {
+    virtualNetworkName: 'dep-<<namePrefix>>-vnet-${serviceShort}'
+    managedIdentityName: 'dep-<<namePrefix>>-msi-${serviceShort}'
+    eventHubName: 'dt-${uniqueString(serviceShort)}-evh-01'
+    eventHubNamespaceName: 'dt-${uniqueString(serviceShort)}-evhns-01'
+    serviceBusName: 'dt-${uniqueString(serviceShort)}-sb-01'
+    serviceBusTopicName: 'dt-${uniqueString(serviceShort)}-sbtp-01'
```
Comment (AlexanderSehr) [inline modules/Microsoft.DigitalTwins/digitalTwinsInstances/.test/common/deploy.test.bicep:40]: Question: Do the diverse topic names need to be unique? If not, we could also just hardcode them in the dependencies file like we do for example for the virtual network subnet - effectively reducing the amount of parameters

File: modules/Microsoft.DigitalTwins/digitalTwinsInstances/.test/common/dependencies.bicep [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+
+output eventhubNamespaceName string = eventHubNamespace.name
+
+output eventHubId string = eventHub.id
+
+output eventhubName string = eventHub.name
+
+output serviceBusName string = serviceBus.name
+
+output serviceBusTopicName string = serviceBusTopic.name
```
Comment (AlexanderSehr) [inline modules/Microsoft.DigitalTwins/digitalTwinsInstances/.test/common/dependencies.bicep:154]:
```suggestion
output serviceBusTopicName string = serviceBus::topic.name
```

File: modules/Microsoft.DigitalTwins/digitalTwinsInstances/.test/common/dependencies.bicep [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+
+output eventGridDomainName string = eventGridDomain.name
+
+output eventGridTopicName string = eventGridTopic.name
```
Comment (AlexanderSehr) [inline modules/Microsoft.DigitalTwins/digitalTwinsInstances/.test/common/dependencies.bicep:158]:
```suggestion
output eventGridTopicName string = eventGrid::topic.name
```

File: modules/Microsoft.DigitalTwins/digitalTwinsInstances/.test/common/dependencies.bicep [hunk 14]
```diff
@@ -0,0 +1,166 @@
...
+
+output eventGridEndpoint string = eventGridDomain.properties.endpoint
+
+output eventGridTopicId string = eventGridTopic.id
```
Comment (AlexanderSehr) [inline modules/Microsoft.DigitalTwins/digitalTwinsInstances/.test/common/dependencies.bicep:162]:
```suggestion
output eventGridTopicResourceId string = eventGrid::topic.id
```

File: modules/Microsoft.DigitalTwins/digitalTwinsInstances/.test/common/dependencies.bicep (hunk 13 above)
Comment (AlexanderSehr) [inline modules/Microsoft.DigitalTwins/digitalTwinsInstances/.test/common/dependencies.bicep:158]: Do we need these outputs actually? I see some of them aren't actually used. If not, please remove any which are redundant.

File: modules/Microsoft.DigitalTwins/digitalTwinsInstances/endpoints-eventGrid/deploy.bicep [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+param name string = 'EventGridEndpoint'
+
+@description('Conditional. The name of the parent Digital Twin Instance resource. Required if the template is used in a standalone deployment.')
+param digitalTwinInstanceName string
+
+@description('Required. EventGrid Topic Endpoint.')
+param topicEndpoint string
+
+@description('Required. Event Grid Resource Id.')
+param eventGridDomainId string
```
Comment (AlexanderSehr) [inline modules/Microsoft.DigitalTwins/digitalTwinsInstances/endpoints-eventGrid/deploy.bicep:11]:
```suggestion
param eventGridDomainResourceId string
```

File: modules/Microsoft.DigitalTwins/digitalTwinsInstances/endpoints-eventGrid/deploy.bicep [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+@description('Optional. Dead letter storage URL for identity-based authentication.')
+param deadLetterUri string = ''
+
+@description('Optional. Enable telemetry via the Customer Usage Attribution ID (GUID).')
+param enableDefaultTelemetry bool = true
+
+var eventGridAccessKey1 = listkeys(eventGridDomainId, '2022-06-15').key1
+
+var eventGridAccessKey2 = listkeys(eventGridDomainId, '2022-06-15').key2
+
```
Comment (AlexanderSehr) [inline modules/Microsoft.DigitalTwins/digitalTwinsInstances/endpoints-eventGrid/deploy.bicep:26]:
```suggestion
```

I'd recommend to fetch them directly in the parameters block as they are not used anywhere else.

File: modules/Microsoft.DigitalTwins/digitalTwinsInstances/endpoints-eventGrid/readme.md [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+<details>
+
+<summary>Parameter JSON format</summary>
+
+```json
+"eventGridEndpoint": {
+    "value": {
+        "name": "<Endpoint Name>",
+        "accessKey1": "",
+        "accessKey2": "",
```
Comment (AlexanderSehr) [inline modules/Microsoft.DigitalTwins/digitalTwinsInstances/endpoints-eventGrid/readme.md:60]: I guess now that we fetch the access key from the `eventGridDomainResourceId`, we can update both this and the bicep example.

File: modules/Microsoft.DigitalTwins/digitalTwinsInstances/endpoints-eventHub/readme.md [hunk 18]
```diff
@@ -0,0 +1,103 @@
...
+}
+  ```
+
+</details>
+<p>
+
+
+
+
+
```
Comment (AlexanderSehr) [inline modules/Microsoft.DigitalTwins/digitalTwinsInstances/endpoints-eventHub/readme.md:92]:
```suggestion

```

File: modules/Microsoft.DigitalTwins/digitalTwinsInstances/endpoints-eventGrid/readme.md (hunk 17 above)
Comment (AlexanderSehr) [reply modules/Microsoft.DigitalTwins/digitalTwinsInstances/endpoints-eventGrid/readme.md:60]: Also this example should only exist in the parent readme.

File: modules/Microsoft.DigitalTwins/digitalTwinsInstances/endpoints-eventHub/readme.md [hunk 19]
```diff
@@ -0,0 +1,103 @@
...
+| `userAssignedIdentity` | string | `''` | The ID to assign to the resource. |
+
+### Parameter Usage: `eventhubEndpoint`
+
+<details>
+
+<summary>Parameter JSON format</summary>
+
+```json
+"eventhubEndpoint": {
```
Comment (AlexanderSehr) [inline modules/Microsoft.DigitalTwins/digitalTwinsInstances/endpoints-eventHub/readme.md:52]: This example should only exist in the parent readme.

File: modules/Microsoft.DigitalTwins/digitalTwinsInstances/endpoints-serviceBus/readme.md [hunk 20]
```diff
@@ -0,0 +1,102 @@
...
+
+
+### Parameter Usage: `serviceBusEndpoint`
+
+<details>
+
+<summary>Parameter JSON format</summary>
+
+```json
+"serviceBusEndpoint": {
```
Comment (AlexanderSehr) [inline modules/Microsoft.DigitalTwins/digitalTwinsInstances/endpoints-serviceBus/readme.md:53]: This example should only exist in the parent readme.

File: modules/Microsoft.DigitalTwins/digitalTwinsInstances/readme.md [hunk 21]
```diff
@@ -0,0 +1,611 @@
...
+<p>
+
+### Parameter Usage: `eventhubEndpoint`
+
+<details>
+
+<summary>Parameter JSON format</summary>
+
+```json
+"eventhubEndpoint": {
```
Comment (AlexanderSehr) [inline modules/Microsoft.DigitalTwins/digitalTwinsInstances/readme.md:299]: Please update this and the other examples based on the latest comments in the individual child-module readmes

File: modules/Microsoft.DigitalTwins/digitalTwinsInstances/.test/common/dependencies.bicep (hunk 10 above)
Comment (ryanmstephens) [reply modules/Microsoft.DigitalTwins/digitalTwinsInstances/.test/common/dependencies.bicep:61]: Thanks just verified that its [privatelink.digitaltwins.azure.net](privatelink.digitaltwins.azure.net). The target sub resource is "API". I will update this.

File: modules/Microsoft.DigitalTwins/digitalTwinsInstances/.test/common/deploy.test.bicep (hunk 11 above)
Comment (ryanmstephens) [reply modules/Microsoft.DigitalTwins/digitalTwinsInstances/.test/common/deploy.test.bicep:40]: Correct it doesn't require a unique name. I will hard code the value into dependency template.

File: modules/Microsoft.DigitalTwins/digitalTwinsInstances/.test/common/dependencies.bicep (hunk 13 above)
Comment (ryanmstephens) [reply modules/Microsoft.DigitalTwins/digitalTwinsInstances/.test/common/dependencies.bicep]: Agreed - these were left in from testing. I will remove them.

File: modules/Microsoft.DigitalTwins/digitalTwinsInstances/.test/common/deploy.test.bicep (hunk 11 above)
Comment (ryanmstephens) [reply modules/Microsoft.DigitalTwins/digitalTwinsInstances/.test/common/deploy.test.bicep:40]: Renamed

File: modules/Microsoft.DigitalTwins/digitalTwinsInstances/endpoints-eventGrid/readme.md (hunk 17 above)
Comment (ryanmstephens) [reply modules/Microsoft.DigitalTwins/digitalTwinsInstances/endpoints-eventGrid/readme.md]: Removed from child readme files. The example already existed in the parent readmes as well.

File: modules/Microsoft.DigitalTwins/digitalTwinsInstances/endpoints-eventHub/readme.md (hunk 19 above)
Comment (ryanmstephens) [reply modules/Microsoft.DigitalTwins/digitalTwinsInstances/endpoints-eventHub/readme.md]: Removed from child readme file. It was duplicated in the parent readme file, so it was already in the parent readme file.

File: modules/Microsoft.DigitalTwins/digitalTwinsInstances/endpoints-serviceBus/deploy.bicep [hunk 22]
```diff
@@ -0,0 +1,90 @@
...
+
+@description('Optional. The URL of the ServiceBus namespace for identity-based authentication. It must include the protocol \'sb://\'.')
+param endpointUri string = ''
+
+@description('Optional. The ServiceBus Topic name for identity-based authentication.')
+param entityPath string = ''
+
+@description('Optional. PrimaryConnectionString of the endpoint for key-based authentication. Will be obfuscated during read.')
+@secure()
+param primaryConnectionString string = ''
```
[the remark this answers is not in the document]
Comment (ryanmstephens) [reply modules/Microsoft.DigitalTwins/digitalTwinsInstances/endpoints-serviceBus/deploy.bicep:29]: Service Bus works for both identity and key based auth. Event Grid is the only endpoint that only supports key based auth.

File: modules/Microsoft.DigitalTwins/digitalTwinsInstances/endpoints-serviceBus/readme.md (hunk 20 above)
Comment (ryanmstephens) [reply modules/Microsoft.DigitalTwins/digitalTwinsInstances/endpoints-serviceBus/readme.md]: Removed from child. It was duplicated in the parent so already existed in that readme file.

File: modules/Microsoft.DigitalTwins/digitalTwinsInstances/endpoints-serviceBus/deploy.bicep (hunk 22 above)
[the remark this answers is not in the document]
Comment (ryanmstephens) [reply modules/Microsoft.DigitalTwins/digitalTwinsInstances/endpoints-serviceBus/deploy.bicep:29]: Updated description to Service Bus and Event Hub to conditional parameter. Additionally, these have been added to readme file conditional parameter table.

File: modules/Microsoft.DigitalTwins/digitalTwinsInstances/.test/common/dependencies.bicep (hunk 6 above)
Comment (ryanmstephens) [reply modules/Microsoft.DigitalTwins/digitalTwinsInstances/.test/common/dependencies.bicep]: Added comment to explain the role id

File: modules/Microsoft.DigitalTwins/digitalTwinsInstances/.test/common/dependencies.bicep (hunk 9 above)
Comment (ryanmstephens) [reply modules/Microsoft.DigitalTwins/digitalTwinsInstances/.test/common/dependencies.bicep]: Add comment to explain role id

File: modules/Microsoft.DigitalTwins/digitalTwinsInstances/.test/common/dependencies.bicep [hunk 23]
```diff
@@ -107,60 +103,55 @@ resource eventGridDomain 'Microsoft.EventGrid/domains@2022-06-15' = {
...
+output managedIdentityPrincipalResourceId string = managedIdentity.properties.principalId
 
 @description('The resource ID of the created Private DNS Zone.')
-output privateDNSResourceId string = privateDNSZone.id
+output privateDNSResourceResourceId string = privateDNSZone.id
 
 output eventhubNamespaceName string = eventHubNamespace.name
 
-output eventHubId string = eventHub.id
+output eventHubResourceResourceId string = eventHubNamespace::eventHub.id
```
Comment (AlexanderSehr) [inline modules/Microsoft.DigitalTwins/digitalTwinsInstances/.test/common/dependencies.bicep:143]: It appears that you accidently duplicated `Resource` in some instances like above:
`eventHubResourceResourceId` should be `eventHubResourceId`.

Please replace all these occurences. Should be as easy as a Search & Replace of all `ResourceResourceId` with `ResourceId`.

File: modules/Microsoft.DigitalTwins/digitalTwinsInstances/.test/common/dependencies.bicep [hunk 24]
```diff
@@ -0,0 +1,157 @@
...
+@description('The resource ID of the created Virtual Network Subnet.')
+output subnetResourceId string = virtualNetwork.properties.subnets[0].id
+
+@description('The principal ID of the created Managed Identity.')
+output managedIdentityPrincipalResourceId string = managedIdentity.properties.principalId
+
+@description('The resource ID of the created Private DNS Zone.')
+output privateDNSResourceId string = privateDNSZone.id
+
+output eventhubNamespaceName string = eventHubNamespace.name
```
Comment (AlexanderSehr) [inline modules/Microsoft.DigitalTwins/digitalTwinsInstances/.test/common/dependencies.bicep:141]: Please add the remaining metadata

File: modules/Microsoft.DigitalTwins/digitalTwinsInstances/.test/common/dependencies.bicep [hunk 25]
```diff
@@ -0,0 +1,157 @@
...
+
+output eventHubResourceId string = eventHubNamespace::eventHub.id
+
+output eventhubName string = eventHubNamespace::eventHub.name
+
+output serviceBusName string = serviceBus.name
+
+output serviceBusTopicName string = serviceBus::topic.name
+
+output eventGridEndpoint string = eventGridDomain.properties.endpoint
```
Comment (AlexanderSehr) [inline modules/Microsoft.DigitalTwins/digitalTwinsInstances/.test/common/dependencies.bicep:151]: Is that the endpoint name?

**Review by AlexanderSehr — APPROVED** [review]

**Pull request merged**